### PR TITLE
[coqide] Revert behavior changes in offset location

### PR DESCRIPTION
**Pull request by ejgallego**
This partially reverts changes to `coqidetop` and `coqide` regarding how locations are done, corresponding PRs are #16978 and #17026

Hack present in 8.16 is still present, but pushed to CoqIDE layer.

This PR includes the reversions to ensure this is the behavior we want, then a refix for #15087 but while preserving the hack in the CoqIDE layer.

Reopens #16987 , which will have to stay open IIUC.

### CODE REVIEW

File: ide/coqide/idetop.ml
Comment (ejgallego) [reply ide/coqide/idetop.ml:96]: Actually I have no idea why fix_loc was necessary in the first place, and it created a serious bug in the first place (https://github.com/coq/coq/issues/15087) . This patch just moves fix_loc to be coqide specific.

Comment (jfehrle) [reply ide/coqide/idetop.ml:96]: > Actually I have no idea why fix_loc was necessary in the first place, and it created a serious bug (https://github.com/coq/coq/issues/15087).

`fix_loc` is "necessary" because #16978 didn't fix #15087.

Here's an excerpt from `Idetop.add`, lines 103-105.  The first line includes `bp=off` to initialize the parser, **which is ignored**.  That's the bug of #15087.  Adding a printf in `fix_loc` shows that bp/ep are relative to the beginning of the statement rather than the beginning of the buffer ("absolute").

What you have in this PR works for CoqIDE, but it seems likely you will need a similar `fix_loc` to support the debugger in LSP, too--at least if you want breakpoints that are after the first sentence to work.  Better to fix this in the lexer/parser than in both XML and LSP.

You may want to revert the `fix_loc` changes in this PR.  Fixing the lexer/parser can be done later.  I won't object if you and @Zimmi48 want to merge this as-is.

```
  let loc = { (initial ToplevelInput) with bp=off; line_nb } in
  let fix_loc = fix_loc ~off in
  let pa = Pcoq.Parsable.make ~loc ~fix_loc (Gramlib.Stream.of_string s) in
```

```
bp = 4 ep = 8 line_nb = 5 off = 119
bp = 9 ep = 13 line_nb = 5 off = 119
bp = 13 ep = 14 line_nb = 5 off = 119
bp = 2 ep = 7 line_nb = 6 off = 133
bp = 7 ep = 8 line_nb = 6 off = 133
bp = 2 ep = 6 line_nb = 7 off = 141
bp = 6 ep = 7 line_nb = 7 off = 141
bp = 11 ep = 14 line_nb = 8 off = 148
bp = 14 ep = 15 line_nb = 8 off = 148
bp = 4 ep = 8 line_nb = 10 off = 163
bp = 9 ep = 13 line_nb = 10 off = 163
```

Comment (ejgallego) [reply ide/coqide/idetop.ml:96]: This is not how Coq's lexer and parser work, bp will be set from `Stream.count ()`, so in the above case you need to properly call `Stream.of_string ~offset`  with the right offset. Please read #16978 again.

The above code is indeed nonsensical but equivalent to what was in 8.16, the goal is to preserve the exact behavior as the hack removed in the fix set some parts of the location at random.

There is no fixing to do w.r.t. parsing resumption (which has been working since a long time in LSP).

Comment (jfehrle) [reply ide/coqide/idetop.ml:96]: Then there's no point in creating `fix_loc` at all.  The offset should be passed in the in the call setting up the stream.  The code should be like this:

```
let add ((((s,eid),(sid,verbose)),offset),(line_nb,bol_pos)) =
  let doc = get_doc () in
  (* note: this won't yield correct values for bol_pos_last,
     but the debugger doesn't use that *)
  let loc = Loc.{ (initial ToplevelInput) with line_nb } in
  let pa = Pcoq.Parsable.make ~loc (Gramlib.Stream.of_string ~offset s) in
```

Of course, the `offset` option didn't exist when I was working on this.

**Review by jfehrle — APPROVED** [review]
I'll remove `fix_loc` in a separate PR.

I will merge on Tuesday if there are no further comments.

File: ide/coqide/idetop.ml
Comment (ejgallego) [reply ide/coqide/idetop.ml:96]: This is what master does.

**Comment by ejgallego** [comment]
It seems to me, based on the discussions that were held, that in the end we should not merge this PR and instead go ahead with the right fix #17391 

I am sorry for the sequence of events / PRs that weren't supposed to happen like this, let me recap:
- in the Coq Call we agreed that we would revert CoqIDE to 8.16 behavior
- I committed to do the PR (this one) and the 8.17 backport (already merged)
- I did however a call for testing, that @herbelin promptly answered, he came to my office and after some work and code / behavior assessment we did #17391 
- I wasn't planning to do this, but in the end even if unplanned, made sense, however as it is an impromptu deviation of the original plan things got a bit messy.

But now it is clear that adding `fix_loc` to `master` is not something we want, so I propose we close this PR and continue the plan discussed, that will yield a `master` that is finally almost totally ordered over previous versions with bugs (with the order being "less buggy")

As for 8.17 IMHO it can go with either the revert or the code in `master`, given the current status I'd stay with the revert for now.

There are a couple of minor points in the backport PR that I will address there.

**Comment by ejgallego** [comment]
Superseeded by #17391

**Comment by jfehrle** [comment]
> Superseeded by #17391

Did you try testing that?  That version is much worse.

**Comment by ejgallego** [comment]
> That version is much worse.

In what is worse? Yes, we did test that and it worked fine.

**Comment by ejgallego** [comment]
Oh sorry I got the wrong PR, I did mean #17382 , github completion tricked me.

**Pull request closed without merging**
